pages/api/teams/[teamId]/invite.ts

```diff
@@ -0,0 +1,148 @@
+import { randomUUID } from "crypto";
+import { getServerSession } from "next-auth";
+import { authOptions } from "../../auth/[...nextauth]";
+import { NextApiRequest, NextApiResponse } from "next";
+import prisma from "@/lib/prisma";
+import { CustomUser } from "@/lib/types";
+import { errorHanlder } from "@/lib/errorHandler";
+import { sendTeammateInviteEmail } from "@/lib/emails/send-teammate-invite";
+
+export default async function handle(
+  req: NextApiRequest,
+  res: NextApiResponse
+) {
+  if (req.method === "POST") {
+    // POST /api/teams/:teamId/invite
+    const session = await getServerSession(req, res, authOptions);
+    if (!session) {
+      return res.status(401).end("Unauhorized");
+    }
+
+    const { teamId } = req.query as { teamId: string };
+
+    const { email } = req.body;
+
+    if (!email) {
+      return res.status(400).json("Email is missing in request body");
+    }
+
+    try {
+      const team = await prisma.team.findUnique({
+        where: {
+          id: teamId,
+        },
+        include: {
+          users: {
+            select: {
+              userId: true,
+              role: true,
+            },
+          },
+        },
+      });
+
+      // check that the user is admin of the team, otherwise return 403
+      const teamUsers = team?.users;
+      const isUserAdmin = teamUsers?.some(
+        (user) =>
+          user.role === "ADMIN" &&
+          user.userId === (session.user as CustomUser).id
+      );
+      if (!isUserAdmin) {
+        return res.status(403).json("Only admins can send the invitation!");
+      }
+
+      const token = randomUUID().toString();
+
+      const expiresAt = new Date();
+      expiresAt.setHours(expiresAt.getHours() + 24); // invitation expires in 24 hour
+      const invitation = await prisma.invitation.create({
+        data: {
+          email,
+          token,
+          expires: expiresAt,
+          teamId,
+        },
+      });
+
+      // send invite email
+      const sender = session.user as CustomUser;
+
+      sendTeammateInviteEmail({
+        senderName: sender.name || "",
+        senderEmail: sender.email || "",
+        teamName: team?.name || "",
+        teamId: team?.id || "",
+        token,
+        to: email,
+      });
+
+      return res.status(200).json("Invitation send!");
+    } catch (error) {
+      errorHanlder(error, res);
+    }
+  } else if (req.method === "GET") {
+    // GET /api/teams/:teamId/invite
+    const session = await getServerSession(req, res, authOptions);
+
+    const { token, teamId, email } = req.query as {
+      token: string;
+      teamId: string;
+      email: string;
+    };
+
+    if (!session) {
+      res.redirect(`/login?next=/api/teams/${teamId}/invite?token=${token}`);
+    }
+
+    if (email !== session?.user?.email) {
+      return res.status(400).json("You are not invited to the team");
+    }
```

Comment:
This throws an error on the invite flow if the user is not logged in. 

Becausue in Line 95, it's redirecting to the api route without the `email` query param. 

We should not read the `email` from the url param at all. Only read it through the `token` / `invitation` model to prevent someone manipulating the url with a different email and then able to join the team.